src/fluree/db/connection.cljc

```diff
@@ -445,6 +446,88 @@
     (load-ledger-address conn alias-or-address)
     (load-ledger-alias conn alias-or-address)))
 
+(defn drop-commit-artifacts
+  [{:keys [commit-catalog] :as _conn} latest-commit]
+  (let [error-ch  (async/chan)
+        commit-ch (commit-storage/trace-commits commit-catalog latest-commit 0 error-ch)]
+    (go-loop []
+      (when-let [[commit _] (<! commit-ch)]
+        (let [txn-address         (util/get-first-value commit const/iri-txn)
+              commit-address      (util/get-first-value commit const/iri-address)
+              data-address        (-> (util/get-first commit const/iri-data)
+                                      (util/get-first-value const/iri-address))]
+          (log/debug "Dropping commit" (-> (util/get-first commit const/iri-data)
+                                           (util/get-first-value const/iri-fluree-t)))
+          (when data-address
+            (log/debug "Deleting data" data-address)
+            (storage/delete commit-catalog data-address))
+          (when commit-address
+            (log/debug "Deleting commit" commit-address)
+            (storage/delete commit-catalog commit-address))
+          (when txn-address
+            (log/debug "Deleting txn" txn-address)
+            (storage/delete commit-catalog txn-address))
+          (recur))))))
+
+(defn drop-index-nodes
+  "Build up a list of node addresses in leaf->root order, then delete them."
+  [storage node-address]
+  (go-try
+    (loop [[address & r] [node-address]
+           addresses     (list)]
+      (if address
+        (if-let [children (->> (:children (<? (storage/read-json storage address true)))
+                               (mapv :id))]
+          (recur (into r children) (conj addresses address))
+          (recur r (conj addresses address)))
+
+        (doseq [address addresses]
+          (log/debug "Dropping node" address)
+          (storage/delete storage address))))
+    :nodes-dropped))
+
+(defn drop-index-artifacts
+  [{:keys [index-catalog] :as _conn} latest-commit]
+  (go-try
+    (let [storage       (:storage index-catalog)
+          index-address (some-> (util/get-first latest-commit const/iri-index)
+                                (util/get-first-value const/iri-address))]
```

Comment:
I don't think this will capture a virtual graph index (BM25), but hard to tell from first glance.

Reply:
I think we should move forward with this PR and deal with the Virtual Graph stuff later - it is a moving target regardless. @zonotope do you agree, if so please review as well - thanks!